stdlib/public/Reflection/Sources/_Runtime/ImageInspection.swift

```diff
@@ -0,0 +1,97 @@
+//===----------------------------------------------------------------------===//
+//
+// This source file is part of the Swift.org open source project
+//
+// Copyright (c) 2023 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+//
+//===----------------------------------------------------------------------===//
+
+import Swift
+
+@available(SwiftStdlib 5.9, *)
+@frozen
+public enum ImageInspection {}
+
+@_silgen_name("snapshot")
+func snapshot()
+
+//===----------------------------------------------------------------------===//
+// Runtime Attributes
+//===----------------------------------------------------------------------===//
+
+@available(SwiftStdlib 5.9, *)
+var attributeCache: Lock<[ContextDescriptor: [UnsafeRawPointer]]> = .create(
+  with: [:]
+)
+
+@available(SwiftStdlib 5.9, *)
+extension ImageInspection {
+  @available(SwiftStdlib 5.9, *)
+  public static func withAttributeCache<T>(
+    _ body: @Sendable ([ContextDescriptor: [UnsafeRawPointer]]) throws -> T
+  ) rethrows -> T {
+#if !(os(macOS) || os(iOS) || os(tvOS) || os(watchOS))
+    attributeCache.withLock {
+      $0.removeAll(keepingCapacity: true)
+    }
+#endif
+
+    snapshot()
+
+    return try attributeCache.withLock {
+      try body($0)
+    }
+  }
+}
+
+@available(SwiftStdlib 5.9, *)
+@_cdecl("registerAttributes")
```

Comment:
This puts the function in the global C namespace with the name "registerAttributes". Did you want "swift_registerAttributes" instead?

Reply:
Internally, yes, but these symbols are not exported. This is only an issue if a dependency of this library declares some symbol named this. I can change this if you want.

Reply:
See my other comments. I think we can refactor this code to not need this function at all.

I'd be happy to offer an additional commit to this PR, or to open a separate PR, to show what I mean. :)